internal/events/subscription_manager.go

```diff
@@ -275,47 +289,84 @@ func (sm *subscriptionManager) parseSubscriptionDef(ctx context.Context, subDef
 		}
 	}
 
-	var tagFilter *regexp.Regexp
-	if filter.Tag != "" {
-		tagFilter, err = regexp.Compile(filter.Tag)
-		if err != nil {
-			return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.tag", filter.Tag)
-		}
+	sub = &subscription{
+		dispatcherElection: make(chan bool, 1),
+		definition:         subDef,
+		eventMatcher:       eventFilter,
 	}
 
-	var groupFilter *regexp.Regexp
-	if filter.Group != "" {
-		groupFilter, err = regexp.Compile(filter.Group)
-		if err != nil {
-			return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.group", filter.Group)
+	if (filter.Message != fftypes.MessageFilter{}) {
+		var tagFilter *regexp.Regexp
+		if filter.Message.Tag != "" {
+			tagFilter, err = regexp.Compile(filter.Message.Tag)
+			if err != nil {
+				return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.Message.tag", filter.Message.Tag)
+			}
 		}
-	}
 
-	var topicsFilter *regexp.Regexp
-	if filter.Topics != "" {
-		topicsFilter, err = regexp.Compile(filter.Topics)
-		if err != nil {
-			return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.topics", filter.Topics)
+		var groupFilter *regexp.Regexp
+		if filter.Message.Group != "" {
+			groupFilter, err = regexp.Compile(filter.Message.Group)
+			if err != nil {
+				return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.Message.group", filter.Message.Group)
+			}
+		}
+
+		var topicsFilter *regexp.Regexp
+		if filter.Message.Topics != "" {
+			topicsFilter, err = regexp.Compile(filter.Message.Topics)
+			if err != nil {
+				return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.Message.topics", filter.Message.Topics)
+			}
+		}
+
+		var authorFilter *regexp.Regexp
+		if filter.Message.Author != "" {
+			authorFilter, err = regexp.Compile(filter.Message.Author)
+			if err != nil {
+				return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.Message.author", filter.Message.Author)
+			}
+		}
+
+		msg := &messageFilter{
+			tagFilter:    tagFilter,
+			groupFilter:  groupFilter,
+			topicsFilter: topicsFilter,
+			authorFilter: authorFilter,
 		}
+		sub.messageFilter = msg
 	}
 
-	var authorFilter *regexp.Regexp
-	if filter.Author != "" {
-		authorFilter, err = regexp.Compile(filter.Author)
-		if err != nil {
-			return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.author", filter.Author)
+	if (filter.BlockchainEvent != fftypes.BlockchainEventFilter{}) {
+		var nameFilter *regexp.Regexp
+		if filter.BlockchainEvent.Name != "" {
+			nameFilter, err = regexp.Compile(filter.BlockchainEvent.Name)
+			if err != nil {
+				return nil, i18n.WrapError(ctx, err, i18n.MsgRegexpCompileFailed, "filter.BlockchainEvent.name", filter.BlockchainEvent.Name)
+			}
+		}
+
+		bf := &blockchainFilter{
+			nameFilter: nameFilter,
 		}
+		sub.blockchainFilter = bf
```

Comment:
This is good, and I would keep this, but I'd also like to see filtering by `ContractListener` here as well. That's the most specific thing we can filter by for BlockchainEvents. Just filtering on name is not bad, but for example, if you had multiple ERC20 contracts both emitting `Transfer` events, this filter could pick up both of them.